src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -271,12 +418,24 @@ paAddElem k e PageAcc{..}
     entryToCrumb Mupdate{}        = 1
     entryToCrumb Delete{}         = 2
 
-paSingleton :: Int -> SerialisedKey -> Entry RawValue BlobRef -> PageAcc
-paSingleton rfp k e = fromMaybe (error err) $ paAddElem k e (paEmpty rfp)
+paSingleton :: SerialisedKey -> Entry RawValue BlobRef -> PageAcc
+paSingleton k e = fromMaybe (error err) $
+    -- range-finder precision only matters when the page is non-empty, so any
+    -- value would suffice, so we pick 0.
+    paAddElem 0 k e paEmpty
   where
     err = "Failed to add k/op pair to an empty page, but this should have \
           \worked! Are you sure the implementation of paAddElem is correct?"
 
+unsafeMkAppend :: PageAcc -> Index.Append
+unsafeMkAppend p = case unStricterList $ pageKeys p of
+    []                     -> error "mkAppend: empty list"
+    [k] | numBytes <= 4096 -> Index.AppendSinglePage k k
+        | otherwise        -> Index.AppendMultiPage  k (fromIntegral $ (numBytes - 1) `quot` 4096)
+    ks                     -> Index.AppendSinglePage (last ks) (head ks)
```

Comment:
The `last` seems like it could be avoided, but it probably doesn't matter too much at this point.